openmls/src/messages/proposals_in.rs

```diff
@@ -0,0 +1,324 @@
+// is complete.
+
+impl From<AddProposalIn> for crate::messages::proposals::AddProposal {
+    fn from(value: AddProposalIn) -> Self {
+        Self {
+            key_package: value.key_package,
+        }
+    }
+}
+
+impl From<crate::messages::proposals::AddProposal> for AddProposalIn {
+    fn from(value: crate::messages::proposals::AddProposal) -> Self {
+        Self {
+            key_package: value.key_package,
+        }
+    }
+}
+
+impl From<UpdateProposalIn> for crate::messages::proposals::UpdateProposal {
+    fn from(value: UpdateProposalIn) -> Self {
+        Self {
+            leaf_node: value.leaf_node,
+        }
+    }
+}
+
+impl From<crate::messages::proposals::UpdateProposal> for UpdateProposalIn {
+    fn from(value: crate::messages::proposals::UpdateProposal) -> Self {
+        Self {
+            leaf_node: value.leaf_node,
+        }
+    }
+}
+
+impl From<ProposalIn> for crate::messages::proposals::Proposal {
+    fn from(proposal: ProposalIn) -> Self {
+        match proposal {
+            ProposalIn::Add(add) => Self::Add(add.into()),
+            ProposalIn::Update(update) => Self::Update(update.into()),
+            ProposalIn::Remove(remove) => Self::Remove(remove),
+            ProposalIn::PreSharedKey(psk) => Self::PreSharedKey(psk),
+            ProposalIn::ReInit(reinit) => Self::ReInit(reinit),
+            ProposalIn::ExternalInit(external_init) => Self::ExternalInit(external_init),
+            ProposalIn::GroupContextExtensions(group_context_extension) => {
+                Self::GroupContextExtensions(group_context_extension)
+            }
+            ProposalIn::AppAck(app_ack) => Self::AppAck(app_ack),
+        }
+    }
+}
+
+impl From<crate::messages::proposals::Proposal> for ProposalIn {
+    fn from(proposal: crate::messages::proposals::Proposal) -> Self {
+        use crate::messages::proposals::Proposal;
+
+        match proposal {
+            Proposal::Add(add) => Self::Add(add.into()),
+            Proposal::Update(update) => Self::Update(update.into()),
+            Proposal::Remove(remove) => Self::Remove(remove),
+            Proposal::PreSharedKey(psk) => Self::PreSharedKey(psk),
+            Proposal::ReInit(reinit) => Self::ReInit(reinit),
+            Proposal::ExternalInit(external_init) => Self::ExternalInit(external_init),
+            Proposal::GroupContextExtensions(group_context_extension) => {
+                Self::GroupContextExtensions(group_context_extension)
+            }
+            Proposal::AppAck(app_ack) => Self::AppAck(app_ack),
+        }
+    }
+}
+
+impl From<ProposalOrRefIn> for crate::messages::proposals::ProposalOrRef {
+    fn from(proposal: ProposalOrRefIn) -> Self {
+        match proposal {
+            ProposalOrRefIn::Proposal(proposal) => Self::Proposal(proposal.into()),
+            ProposalOrRefIn::Reference(reference) => Self::Reference(reference),
+        }
+    }
+}
+
+impl From<crate::messages::proposals::ProposalOrRef> for ProposalOrRefIn {
```

Comment:
Is this From implementation meant to go away at some point? I thought this was the direction where we don't wan to allow conversion.

Reply:
This is the "safe" direction. I guess the goal will still be to get rid of this as well in #1186.